services/device-registry-base/src/test/java/org/eclipse/hono/deviceregistry/service/device/AbstractRegistrationServiceTest.java

```diff
@@ -113,6 +165,255 @@ public void testAssertRegistrationContainsDeviceInfo(final VertxTestContext ctx)
+        when(service.processAssertRegistration(eq(DeviceKey.from(TenantKey.from(Constants.DEFAULT_TENANT), deviceId)), any(Span.class)))
+                .thenReturn(Future.succeededFuture(RegistrationResult.from(HttpURLConnection.HTTP_NOT_FOUND)))
+                .thenReturn(Future.succeededFuture(newRegistrationResult(deviceId, autoProvisioningNotificationSent)));
+
+    }
+
+    private RegistrationResult newRegistrationResult(final String deviceId, final boolean autoProvisioningNotificationSent) {
+        return RegistrationResult.from(HttpURLConnection.HTTP_OK,
+                new JsonObject().put(RegistrationConstants.FIELD_PAYLOAD_DEVICE_ID, deviceId)
+                        .put(RegistrationConstants.FIELD_DATA, new JsonObject()
+                                .put(RegistrationConstants.FIELD_VIA, AbstractRegistrationServiceTest.GATEWAY_ID)
+                                .put(RegistryManagementConstants.FIELD_STATUS, new JsonObject()
+                                        .put(RegistrationConstants.FIELD_AUTO_PROVISIONED, true)
+                                        .put(RegistrationConstants.FIELD_AUTO_PROVISIONING_NOTIFICATION_SENT, autoProvisioningNotificationSent))));
+    }
+
+    /**
+     * Verifies that auto-provisioning still succeeds if the device to be auto-provisioned has already been created
+     * (e.g. by a concurrently running request) and the notification has already been sent.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testAssertRegistrationCanAutoProvisionForAlreadyPresentEdgeDevice(final VertxTestContext ctx) {
+        mockAssertRegistration(GATEWAY_ID, Collections.singletonList(GATEWAY_GROUP_ID), Collections.singletonList(RegistryManagementConstants.AUTHORITY_AUTO_PROVISIONING_ENABLED));
+        mockAssertRegistration(DEVICE_ID, true);
+        mockAddEdgeDevice(HttpURLConnection.HTTP_CONFLICT);
+
+        service.assertRegistration(Constants.DEFAULT_TENANT, DEVICE_ID, GATEWAY_ID, span)
+                .onComplete(ctx.succeeding(result -> {
+                    ctx.verify(() -> {
+                        assertThat(result.getStatus()).isEqualTo(HttpURLConnection.HTTP_OK);
+
+                        verify(service).createDevice(any(), any(), any(), any());
+
+                        verify(sender, never()).sendAndWaitForOutcome(any(), any());
+                        verify(service, never()).updateDevice(eq(Constants.DEFAULT_TENANT), eq(DEVICE_ID), any(), any(), any());
+                    });
+                    ctx.completeNow();
+                }));
+    }
+
+    private void mockAddEdgeDevice(final int httpOk) {
+        when(service.createDevice(any(), any(), any(), any()))
+                .thenAnswer((Answer<Future<OperationResult<Id>>>) invocation -> {
+                    final Optional<String> deviceId = invocation.getArgument(1);
+                    return Future.succeededFuture(OperationResult.ok(httpOk, Id.of(deviceId.get()),
```

Comment:
To make this more robust, I would suggest adding:
`if (!deviceId.isPresent()) return Future.failedFuture("missing device id");`

Reply:
Added...